pivot.py

```diff
@@ -0,0 +1,1723 @@
+            case 'aggregation':
+                for field in aggregation_fields:
+                    if (field['hierarchy'] == hierarchy and
+                            field['name'] != self.field):
+                        field['hierarchy'] -= 1
+                    if field['name'] == self.field:
+                        field['hierarchy'] += 1
+
+        # Sort each list using the "hierarchy" field
+        if x_fields:
+                x_fields = sorted(x_fields, key=lambda x: x['hierarchy'])
+        if y_fields:
+            y_fields = sorted(y_fields, key=lambda x: x['hierarchy'])
+        if aggregation_fields:
+            aggregation_fields = sorted(aggregation_fields, key=lambda x: x['hierarchy'])
+
+        # Transform again to string
+        table_properties = ''
+        for field in x_fields:
+            table_properties += f'name={field["name"]}&position={field["position"]}&hierarchy={field["hierarchy"]}&__'
+        for field in y_fields:
+            table_properties += f'name={field["name"]}&position={field["position"]}&hierarchy={field["hierarchy"]}&__'
+        for field in aggregation_fields:
+            table_properties += f'name={field["name"]}&position={field["position"]}&aggregation={field["aggregation"]}&hierarchy={field["hierarchy"]}&__'
+        return redirect(Index(database_name=self.database_name, table_name=self.table_name, table_properties=table_properties, render=False).url())
+
+class PivotTable(Component):
+    'Pivot Table'
+    __name__ = 'www.pivot_table'
+    _path = None
+
+    database_name = fields.Char('Database Name')
+    table_name = fields.Char('Table Name')
+    grouping_fields = fields.Char('Grouping Fields')
+    result_fields = fields.Char('Result Fields')
+
+    @classmethod
+    def get_url_map(cls):
+        return [
+            Rule('/<string:database_name>/babi/pivot/<string:table_name>/<string:grouping_fields>/<string:result_fields>')
+        ]
+
+    def render(self):
+        grouping_fields = []
+        for grouping_field in self.grouping_fields.split('__'):
+            if grouping_field:
+                groupingfield = GroupingField()
+                for field in grouping_field.split('&'):
+                    if field:
+                        if field.split('=')[0] == 'hierarchy' and field.split('=')[1]:
+                            setattr(groupingfield, field.split('=')[0], int(field.split('=')[1]))
+                        else:
+                            setattr(groupingfield, field.split('=')[0], field.split('=')[1])
+                grouping_fields.append(groupingfield)
+
+        result_fields = []
+        for result_field in self.result_fields.split('__'):
+            if result_field:
+                resultfield = ResultField()
+                for field in result_field.split('&'):
+                    if field:
+                        setattr(resultfield, field.split('=')[0], field.split('=')[1])
+                result_fields.append(resultfield)
```

Comment:
I've seen this (or very similar code) in several functions. I guess it requires an abstraction.